kube-client/src/config/file_loader.rs

```diff
@@ -59,40 +59,42 @@ impl ConfigLoader {
         cluster: Option<&String>,
         user: Option<&String>,
     ) -> Result<Self, KubeconfigError> {
+        let current_context_name = &config.current_context.clone().unwrap_or_default();
         let context_name = if let Some(name) = context {
             name
-        } else if let Some(name) = &config.current_context {
-            name
         } else {
-            return Err(KubeconfigError::CurrentContextNotSet);
+            current_context_name
         };
```

Comment:
Why was this changed?

Reply:
I think we can clean this up to:

```rust
        let context_name = context
            .or_else(|| config.current_context.as_ref())
            .ok_or(KubeconfigError::CurrentContextNotSet)?;
```

Reply:
That was removed because current-context can be null/empty, at least that's how it works on kubectl. Do you think we should not change this in kube-rs?

Reply:
It can be null/empty when parsing, but should return an error when using the config, right?

Reply:
It depends, if we want similar behavior as client-go/kubectl, then an empty current-context is a valid value, because context names can be empty too.

Reply:
cluster and users can have empty names too

Reply:
client-go has [`ConfirmUsable`](https://github.com/kubernetes/kubernetes/blob/57eb5d631ccd615cd161b6da36afc759af004b93/staging/src/k8s.io/client-go/tools/clientcmd/validation.go#L182) validation that's called [when creating a config for REST client](https://github.com/kubernetes/kubernetes/blob/57eb5d631ccd615cd161b6da36afc759af004b93/staging/src/k8s.io/client-go/tools/clientcmd/client_config.go#L133) from kubeconfig. `ConfirmUsable` seems to have validation similar to what we have here and more.

"http://localhost:8080" is from [`ClusterDefaults`](https://github.com/kubernetes/kubernetes/blob/57eb5d631ccd615cd161b6da36afc759af004b93/staging/src/k8s.io/client-go/tools/clientcmd/client_config.go#L44) which has a comment `DEPRECATED`.

Reply:
But do you have a context with an empty name? This for example, is valid and works on kubectl. I haven't really navigated client-go code to understand what is it doing, I'm testing it from a user perspective

```
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: ...
    server: https://kubernetes.docker.internal:6443
  name:
contexts:
- context:
    cluster: 
    user: 
  name: 
current-context:
kind: Config
preferences: {}
users:
- name:
  user:
    client-certificate-data: ...
    client-key-data: ...
```

Reply:
Confirmed kubectl works with a context with empty name when current context is empty.

We should align with client-go (library), and not kubectl (app), though.

---

- client-go's [`Config`](https://github.com/kubernetes/client-go/blob/1ac8d459351e21458fd1041f41e43403eadcbdba/tools/clientcmd/api/v1/types.go#L28) struct. (The link in `file_config.rs` is pointing to the struct after the conversion.)
- [conversion](https://github.com/kubernetes/client-go/blob/7697067af71046b18e03dbda04e01a5bb17f9809/tools/clientcmd/api/v1/conversion.go#L28) from array of named objects to map. Errors on duplicate name. Missing name defaults to empty string because it's Go (nullable uses a pointer type).
- client-go has a validation to [reject empty string context name](https://github.com/kubernetes/kubernetes/blob/57eb5d631ccd615cd161b6da36afc759af004b93/staging/src/k8s.io/client-go/tools/clientcmd/validation.go#L196). kubectl seems to work.
- What to do when there's no match. client-go has validation to reject? kubectl seems to use a deprecated default?

We should probably use default value when missing instead of `Option` to align with client-go's [`Config`](https://github.com/kubernetes/client-go/blob/1ac8d459351e21458fd1041f41e43403eadcbdba/tools/clientcmd/api/v1/types.go#L28) as much as possible. Existing `Option<String>` fields should be fixed if it doesn't match client-go for consistency (not in this PR). We should also follow `omitempty` tags and skip serializing empty.

The problem is allowing `null`. Do we really need to? It just feels weird to me, and I don't think this happens in practice.

Reply:
Maybe kubectl doesn't call ConfirmUsable?

---

I think this topic requires a bit more discussion and I don't want to rush into a solution, but I'd like to get the other changes on this PR merged soon.

As empty/null names are very edge cases scenarios, I'd like to propose that:
- revert changes to "name" fields and keep them as required Strings
- create an issue to discuss how do we want to handle those scenarios and implement that on a subsequent PR

What you think?

Reply:
Yeah, let's do that.

Reply:
It's possible that all this lax parsing is due to supporting config merges. Configs that are invalid in isolation (e.g. no current context) are fine in the merged setting (where another file sets the current-context).

Reply:
OK, I've reverted a bunch of changes so the PR is a lot simpler now. Still open for feedback! Thanks